coral-visualization/src/main/java/com/linkedin/coral/vis/RelNodeVisualizationShuttle.java

```diff
@@ -0,0 +1,182 @@
+import guru.nidi.graphviz.model.Node;
+
+import static guru.nidi.graphviz.model.Factory.*;
+
+
+public class RelNodeVisualizationShuttle extends RelShuttleImpl {
+  private final Map<RelNode, Node> nodeMap;
+
+  public RelNodeVisualizationShuttle() {
+    this.nodeMap = new HashMap<>();
+  }
+
+  @Override
+  public RelNode visit(LogicalAggregate aggregate) {
+    super.visit(aggregate);
+    return aggregate;
+  }
+
+  @Override
+  public RelNode visit(TableScan scan) {
+    super.visit(scan);
+    nodeMap.put(scan, node(String.join(".", scan.getTable().getQualifiedName())));
+    return scan;
+  }
+
+  @Override
+  public RelNode visit(TableFunctionScan scan) {
+    super.visit(scan);
+    nodeMap.put(scan, node(scan.getCall().toString()));
+    return scan;
+  }
+
+  @Override
+  public RelNode visit(LogicalValues values) {
+    super.visit(values);
+    String label = values.getRowType().toString() + " values";
+    nodeMap.put(values, node(values.getRowType().toString() + " values"));
+    return values;
+  }
+
+  @Override
+  public RelNode visit(LogicalFilter filter) {
+    super.visit(filter);
+    Node filterNode = node("Filter: " + filter.getCondition().toString());
+    Node input = nodeMap.get(filter.getInput());
+    nodeMap.put(filter, filterNode.link(input));
+    return filter;
+  }
+
+  @Override
+  public RelNode visit(LogicalProject project) {
+    super.visit(project);
+    Node filterNode = node("Project: " + project.getProjects().toString());
+    Node input = nodeMap.get(project.getInput());
+    nodeMap.put(project, filterNode.link(input));
+    return project;
+  }
+
+  @Override
+  public RelNode visit(LogicalJoin join) {
+    super.visit(join);
+    Node joinNode = node("Join: " + join.getCondition().toString());
+    RelNode leftChild = join.getLeft();
+    RelNode rightChild = join.getRight();
+    String leftLabel = getLabel(leftChild, 0, false);
+    String rightLabel = getLabel(rightChild, leftChild.getRowType().getFieldCount(), false);
+    LinkTarget leftEdge = edge(leftChild, leftLabel);
+    LinkTarget rightEdge = edge(rightChild, rightLabel);
+    nodeMap.put(join, joinNode.link(new LinkTarget[] { leftEdge, rightEdge }));
+    return join;
+  }
+
+  @Override
+  public RelNode visit(LogicalCorrelate correlate) {
+    super.visit(correlate);
+    Node correlateNode = node("Correlate: " + correlate.getCorrelVariable());
+    RelNode leftChild = correlate.getLeft();
+    RelNode rightChild = correlate.getRight();
+    String leftLabel = getLabel(leftChild, 0, true);
+    String rightLabel = getLabel(rightChild, leftChild.getRowType().getFieldCount(), false);
+    LinkTarget leftEdge = edge(leftChild, leftLabel);
+    LinkTarget rightEdge = edge(rightChild, rightLabel);
+    nodeMap.put(correlate, correlateNode.link(new LinkTarget[] { leftEdge, rightEdge }));
+    return correlate;
+  }
+
+  private String getLabel(RelNode target, int startIndex, boolean isCorrelateSide) {
+    String label = "";
```

Comment:
are we able to convert all `$i` representations to column names from rowType?
If so, we can directly use the column names in each node, ex: `Project: [a,b,c,d]` and `hive.test.foo [a,b]` (from scan's rowType) instead of adding text on the edges?